Add utility functions for generating GeoViews plots

BENR0 opened the pull request:
<!-- Please make the PR against the `master` branch. -->

<!-- Describe what your PR does, and why -->
Adds to_geoviews Scene method as show during the Pytroll Developer Week in Darmstadt.
Additionally a new blending function for the MultiScene is added to generate a "timeseries" Scene which in turn can also be used with the geoviews method.

Some documentation is also added. If the data "problem" can be solved Ican also add some Jupyter Notebook examples.

This too can be considered as a work in progress. It's not extensively tested and probably could be improved a lot but maybe is a nice addition to the show method when working in a Notebook.

BENR0 commented on satpy/scene.py:
```diff
@@ -36,7 +36,15 @@
+
+import holoviews as hv
+import geoviews as gv
+import geoviews.feature as gf
```
Ok missed that one. geoviews.feature import only necessary in the Jupyter Notebook so can be removed here.

BENR0 commented on satpy/scene.py:
```diff
@@ -36,7 +36,15 @@
 from pyresample.geometry import AreaDefinition, BaseDefinition
+from pyresample.utils import proj4_str_to_dict
+
+import holoviews as hv
```
Holoviews import also superfluous at this point.

djhoese commented on satpy/scene.py:
```diff
@@ -1023,6 +1031,52 @@ def show(self, dataset_id, overlay=None):
+            # by default select first data variable as display variable
+            vdims = ds.data_vars[ds.data_vars.keys()[0]].name
+
+        if hasattr(ds, "area"):
```
Pyresample `AreaDefinition` objects have a `to_cartopy_crs()` that could be used to replace this.

djhoese replied:
This wouldn't help with `SwathDefinition` objects with plain longitude/latitude arrays, but I'm not familiar with how geoviews would handle that anyway.

djhoese replied:
I would have thought that the object returned by `to_cartopy_crs` is compatible. At least it is compatible with cartopy. There is a cartopy PR that I'm working on to make this more cartopy-friendly, but it has been lacking feedback from the cartopy maintainers due to time constraints from the sound of it.

Regarding swath definitions: ok. I don't know anything about QuadMeshs as far as geoviews thinks of them.

BENR0 replied:
Had another look at it and uses to_cartopy_crs() now.

djhoese commented on satpy/utils.py:
```diff
@@ -261,3 +264,36 @@ def atmospheric_path_length_correction(data, cos_zen, limit=88.):
     return data * corr
+
+
+def scn_to_xrds(scn, datasets = None):
```
In addition to the stickler suggested fix, could this be used to resolve #367? @mraspaud what were your thoughts on that? A `Scene.to_dataset` method might be too confusing.

Either way I think I'd like this utility function to be called `scene_to_xarray_dataset` instead of the abbreviations.

BENR0 replied:
Yes renaming is fine and makes things more clear. I already saw the issue you opened and really like that idea. Also talked to @mraspaud about that in Darmstadt.
It brought us to the point that it woud be nice if the DatasetID, which is limited to the values defined in the named tuple right now, could be more dynamic (that would open up more possibilities e.g. if two platforms with the same channels are in a Scene/MultiScene they should be selectable by other information). Ideally the DatasetID would (also) use the information of the xr.Dataset attributes.

I think @mraspaud had an idea about the dynamic DatasetID in Darmstadt and already started trying something out.

One more thing:
How do I go about your suggested changes? Should I correct them and do another PR or can I somehow edit the PR right here (have not really worked with the github interface up to now).

mraspaud replied:
Yes, we need to work on the datasetID stuff, but it's not needed for this function to work. Look also at what is done in the cf writer, as we transform the scene to a xr.Dataset instance prior to saving to netcdf.

djhoese replied:
My not-so-secret "agenda" is to merge this before the 0.11 release later this month if at all possible. Or at least by early January because I'd like to show some examples in a SatPy talk I have at the AMS Meeting in Phoenix the second week of January.

I think the function is big enough that it doesn't need to be moved to the `Scene` but maybe a `to_xarray_dataset` method could be added to the Scene that imports and calls this utility function? Or is that too redundant/roundabout?

djhoese replied:
Oh and @BENR0, you can push any new commits to this same branch on your fork and the PR will be automatically updated with the changes. Typically github realizes that you changed the code for a particular discussion and will mark the conversation as "resolved" or "outdated".

BENR0 replied:
No you are right just a normal imstande method.

djhoese commented on satpy/scene.py:
```diff
@@ -1023,6 +1031,52 @@ def show(self, dataset_id, overlay=None):
             img.show()
         return img
 
+    def to_geoviews(self, gvtype=gv.Image, datasets=None, kdims=None, vdims=None, dynamic=False):
```
We can't have geoviews as a required dependency on satpy. At least not right now. So if you could move the new imports inside this function that would be preferred.

BENR0 replied:
Yes will do. I wasn't sure how you guys handled that.

djhoese replied:
For conversion methods like this where additional dependencies are needed we tend to do in-method imports. Another alternative, if it ends up being cleaner, is to do the import at the top inside a try/except with a `HAS_GEOVIEWS = True` variable if imported, `False` if not imported, that gets checked inside the function using it.

djhoese commented on satpy/multiscene.py:
```diff
@@ -99,6 +100,17 @@ def stack(datasets):
+        tmp.coords["time"] = pd.DatetimeIndex([ds.attrs["start_time"]])
+        expanded_ds.append(tmp)
+
+    return xr.concat(expanded_ds, dim="time")
```
Does SatPy's Scene have any issues with a DataArray with a third dimension named time? I suppose if it is only used for the geoviews time series right now it won't be a problem.

BENR0 replied:
Good question. I guess I haven't tested that in depth but I think as far as I know it was no problem. I will have a look at it.

djhoese replied:
I think it could be a problem in `save_datasets` or resampling, but yeah this probably won't be used anywhere except for with geoviews at this point so I'm not too concerned.

djhoese replied:
I'm not concerned about this right now. If someone wants to use time series outside of geoviews then they can implement the functionality at that time. SatPy does not have very good handling of multiple times right now and leaves it to the user to manage multiple orbits/images in most cases so this shouldn't be something you need to be concerned with for this PR.

djhoese commented on doc/source/quickstart.rst:
```diff
@@ -127,11 +127,30 @@ To access the loaded data use the wavelength or name:
+                                                                                                    
+Alternatively if working in a Jupyter notebook the scene can be converted to a `geoviews            
+<http://geo.holoviews.org/index.html>` object using the :meth:`~satpy.scene.Scene.to_geoviews`      
+method.                                                                                             
```
Could you make it a little clearer in this paragraph that `geoviews` (and any other packages) must be installed for this to work?

BENR0 replied:
yes

djhoese commented:
Another thing I thought of, if you are an expert with geoviews, do you have any idea how we could make a geoviews visualization where individual dask array chunks appear in the view when they are calculated? I guess it would be implemented as individual datasets added to the geoviews from a single jupyter cell.

BENR0 commented:
I am by no means a geoviews expert. Actually before the Pytroll Week in Darmstadt this functionality was only some snippets I came up with while "avoiding" other work which was then encouraged to push by @mraspaud. That's why I consider this a starting point.

But regarding your question: Do I understand it correctly; instead of waiting for all the calculations to finish and then showing the whole array as a plot you would like the plot to show and then update that plot with each chuck as soon as it is finished?
That would be very interesting I think since calculation can take some time and people (including myself) are often impatient and want to know whats happening.
I don't know if that is possible but I will investigate.

BENR0 commented:
@djhoese  When I showed this in Darmstadt I used multiple time slots from MSG for the timeseries and data from the Modis cloud mask for the quadmesh plot since I had those at hand but any other data would do. If you have some data stored somewhere which I could then (down)load in the notebook I can use that, otherwise if you tell me where I can store the data I can upload mine (Modis is free so no problem, MSG I don't know, need to check).

BENR0 commented:
Well that really sounds hacky :-). I did a little research and maybe it is possible using holoviews.streams Pipes but I am not sure. I asked the folks at pyViz but have no answer yet.

djhoese commented:
Yeah I talked to the ipywidgets folks about this about 6 months ago. They had different packages that I could try out (ipyleaflet, etc). I am also the maintainer of the vispy project which has a notebook widget for doing webgl displays but this may be too complicated. @mraspaud and I brought this up again on slack last week after ipywidgets (I think) announced a new streaming API for streaming video. I'm not sure what would need to go in to that to make image data (a numpy array) look like a stream.

In the end it would be cool if we could have two types of displays:

1. a single image-like rectangle that gets filled in as chunks get computed
2. a map-like widget (ipyleaflet, geoviews, etc) with coastlines and pan-zoom capabilities that had the individual chunks appear as computed. Same as 1, but on a map.

mraspaud commented on satpy/scene.py:
```diff
@@ -1023,6 +1025,58 @@ def show(self, dataset_id, overlay=None):
+            import geoviews as gv
+            from cartopy import crs
+        except ImportError:
+            warnings.warn("This method needs the geoviews package installed.")
```
Also cartopy could be missing :)

djhoese commented:
@BENR0 How goes the git hell with this one?

BENR0 commented:
@djhoese unfortunately currently I am busy with preparations for a paper presentation so I don't find enough time to take care of the PRs.

I already fixed the superfluous imports which stickler came up with this is due to the change from handling the crs manually vs using the to_cartopy_crs method. The wishlist issue is a left over from the git hell ;-) which is also fixed. Just hasen't been pushed yet.

I guess there are still some "lose ends" which in the long run should be improved so this functionality fits in better (geoviews as well as the to_xarray_dataset method) but as you said as long as its only used with geoviews this should be fine for now.

So the things which are left to take care of from my point of view are:
- check documentation again if any changes are needed due to the latest changes
- prepare Jupyter notebook with examples (Probably I could just modify/extend the one I showed in Darmstadt). It would be nice to have some data somewhere online I could use for the notebook. You said you maybe have some data which could be used?

Apart from that maybe you have some more Ideas/ wishes for changes. In that case please let me know.

djhoese commented:
Closing since #567 replaces it.
